### PR TITLE
Prevents that OPENSSL_gmtime incorrectly signals success if gmtime_r …

### PR DESCRIPTION
**Pull request by guidovranken**
…fails, and that struct\* tm result's possibly uninitialized content is used

### CODE REVIEW

**Comment by guidovranken** [comment]
Example of ```gmtime_r``` failing:

```c
#include "crypto/o_time.h"

int main(void)
{
    struct tm _tm;
    time_t t = 0xFF0F3D5300F0C2AC;
    OPENSSL_gmtime(&t, &_tm);
    return 0;
}
```

**Comment by guidovranken** [comment]
With regards to the preceding comment that says ```gmtime_r``` doesn't always return a pointer: if this is still true then it would be worth the effort to at least integrate this patch for systems where it does work and handle other systems with ifdefs. The use of this function extends into OCSP-related functions and it'd be a shame if things malfunctioned due to an uninitialized ```struct tm``` (that could have security ramifications as well if the purported time variables ever leave the system..).

File: crypto/o_time.c [hunk 1]
```diff
@@ -56,7 +56,8 @@ struct tm *OPENSSL_gmtime(const time_t *timer, struct tm *result)
      * should return &data, but doesn't on some systems, so we don't even
      * look at the return value
      */
```
Comment (richsalz) [inline crypto/o_time.c]: ping @guidovranken so that we can merge this :)

**Comment by guidovranken** [comment]
It took me a while, but here it is.

**Comment by levitte** [comment]
I would apply this on all branches

**Comment by richsalz** [comment]
In 1.0.2, b70dc3a; in 1.1.0 07bc93f; and in master, 7c12035.  thanks!

**Comment by yangyangtiantianlonglong** [comment]
hi， @richsalz 
In vxworks 5.5 system, I discover  the return value of gmtime_r is different in kernel mode and user mode.In kernel mode, gmtime_r always returns 0，and return pointer in user mode.Have you tested this?
How can I modify it?  Is there a macro to distinguish between kernel mode and user mode?
https://github.com/openssl/openssl/blob/8c00f267b8df1a8c70eff8198de40aa561299e48/crypto/o_time.c#L40-L43

**Comment by richsalz** [comment]
I am only replying because you at'ed me. :)  To answer your questions: I do not know.

**Comment by yangyangtiantianlonglong** [comment]
@levitte  hi,  
https://github.com/openssl/openssl/pull/1613#issuecomment-516793955 The  issue，Can you help me answer the above question? and I would like to  modify my code  in the vxworks5.5. Do you have a better idea?
```javascript
#elif defined(OPENSSL_SYS_VXWORKS)
     gmtime_r(timer, result); 
     ts = result;
#elif defined(OPENSSL_THREADS) && !defined(OPENSSL_SYS_WIN32) && !defined(OPENSSL_SYS_MACOSX)  
     if (gmtime_r(timer, result) == NULL) 
         return NULL; 
     ts = result; 
#elif defined (OPENSSL_SYS_WINDOWS) && defined(_MSC_VER) && _MSC_VER >= 1400
    if (gmtime_s(result, timer))
        return NULL;
    ts = result;
#else
    ts = gmtime(timer);
    if (ts == NULL)
        return NULL; 
```

**Pull request closed without merging**